projects/RabbitMQ.Client/client/impl/WireFormatting.Write.cs

```diff
@@ -0,0 +1,515 @@
+        public static int WriteBits(ref byte destination, bool val1, bool val2, bool val3)
+        {
+            destination = (byte)(val1.ToByte() + val2.ToByte() * 2 + val3.ToByte() * 4);
+            return 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteBits(ref byte destination, bool val1, bool val2, bool val3, bool val4)
+        {
+            destination = (byte)(val1.ToByte() + val2.ToByte() * 2 + val3.ToByte() * 4 + val4.ToByte() * 8);
+            return 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteBits(ref byte destination, bool val1, bool val2, bool val3, bool val4, bool val5)
+        {
+            destination = (byte)(val1.ToByte() + val2.ToByte() * 2 + val3.ToByte() * 4 + val4.ToByte() * 8 + val5.ToByte() * 16);
+            return 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteBits(ref byte destination,
+            bool val1, bool val2, bool val3, bool val4, bool val5,
+            bool val6, bool val7, bool val8, bool val9, bool val10,
+            bool val11, bool val12, bool val13, bool val14)
+        {
+            destination = (byte)(val8.ToByte() + val7.ToByte() * 2 + val6.ToByte() * 4 + val5.ToByte() * 8 + val4.ToByte() * 16 + val3.ToByte() * 32 + val2.ToByte() * 64 + val1.ToByte() * 128);
+            destination.GetOffset(1) = (byte)(val14.ToByte() * 4 + val13.ToByte() * 8 + val12.ToByte() * 16 + val11.ToByte() * 32 + val10.ToByte() * 64 + val9.ToByte() * 128);
+            return 2;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteLongstr(ref byte destination, ReadOnlySpan<byte> val)
+        {
+            WriteLong(ref destination, (uint)val.Length);
+            Unsafe.CopyBlockUnaligned(ref destination.GetOffset(4), ref val.GetStart(), (uint)val.Length);
+            return 4 + val.Length;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteShort(ref byte destination, ushort val)
+        {
+            NetworkOrderSerializer.WriteUInt16(ref destination, val);
+            return 2;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteShortstr(ref byte destination, ReadOnlySpan<byte> value)
+        {
+            var length = value.Length;
+            if (length <= byte.MaxValue)
+            {
+                destination = (byte)length;
+                Unsafe.CopyBlockUnaligned(ref destination.GetOffset(1), ref value.GetStart(), (uint)value.Length);
+                return length + 1;
+            }
+
+            return ThrowArgumentTooLong(length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteShortstr(ref byte destination, string val, int remainingBytes)
```

Comment:
This is the only place where we pass the remainingBytes. (and have to pass it through all the other methods. Can we get rid of this? 

Why we have it? => We don't do a size check in the GetByteCount method, so we have to do it here. I think it would be simpler to move the check there and throw earlier, then we could also do it with a proper exception

Reply:
I'll take a closer look here.

Reply:
It's needed because the unsafe version of UTF8.GetBytes requires a max number of bytes to write (so it won't write out of bounds). And since we only have a memory address (the ref byte) we need to be able to tell it shouldn't write outside the allowed memory address.

Reply:
You could pass a constant 255 which is max anyway (if the actual size is checked upfront)
But we can also follow up later on this.

Reply:
Removed the `remainingBytes` checks, since they are redundnant anyways since we precalculate the required buffer size when writing.

Reply:
The check also made sure a short stringn was nkt exceeding 255 bytes. Is this still working? Also there was an issue once requesting a proper argument out of range exception with a proper message in this case.

Reply:
Yes, this still works, the UTF8.GetBytes method throws an ArgumentException (instead of ArgumentOutOfRange exception) if we try to write more than 255 bytes.

Reply:
I'm fine with this, but look at #907

Reply:
Yeah, we might want to make an explicit check in the GetByteCount method instead of relying on this happening this late in the serialization phase.

Reply:
I always had the idea of simplifying the get lenght for short string to something like
```
var maxlen = value.Length * MaxBytesPerUTF8Char;
return maxlen < byte.MaxValue ? len : UTF8.GetByteCount(value);

```
Here we could easily add the 255 check as well as optimize small strings by possilbly slightly overallocating the buffer.